### ISSUE TITLE
[RFC] Opt-in mode that makes backward execute in the order predicted by get_current_graph_task_execution_order

### ISSUE THREAD

**Issue by mmjerge** [pytorch/pytorch#198225]
# [RFC] Opt-in mode that makes backward execute in the order predicted by `get_current_graph_task_execution_order`

### Motivation

`saved_tensors_hooks` enables memory-reduction schemes where large saved tensors are replaced by small deltas at pack time and reconstructed at unpack time. When the reconstruction is sequential (state k is rebuilt from state k+1, as in chunked KV-cache training, reversible networks, and similar recompute-from-delta schemes), the unpack hook needs requests to arrive in an order compatible with the reconstruction chain.

Today, that order is unspecified (documented in #198016, discussed in #198014). Real workloads observe out-of-chain-order requests: in awslabs/keys_values#148, a 36-layer / 1500-chunk training run received a request for chunk 1493 before the deltas for chunks 1494 and 1495 had been requested. Library code can defend against this (awslabs/keys_values#149 does, by walking the chain and parking early-reconstructed states), but the defense costs extra copies and considerable complexity that every library implementing this pattern must reinvent.

### Current behavior

The engine already prefers reverse-forward order, but only opportunistically:

- `ReadyQueue` is a priority queue; among *ready* tasks, the one with the highest `sequence_nr` runs first (`CompareNodeTaskTime` in `torch/csrc/autograd/engine.h`).
- A node only becomes ready when all its input gradients have accumulated, so a lower-`sequence_nr` node can run while a higher one is still waiting. This is where chain-order violations come from.
- `sequence_nr` is thread-local (`aten/src/ATen/SequenceNumber.cpp`), so the preference is not globally meaningful for multi-threaded forward.
- Multi-device backward uses one queue per device, so cross-device order is concurrent by construction.

Meanwhile, `torch::autograd::get_current_graph_task_execution_order()` (exposed for checkpoint debugging, requires `set_multithreading_enabled(False)`) already computes a full deterministic schedule: dependency-respecting, preferring higher `sequence_nr`, via a heap over the dependency counts.

### Proposal

An opt-in mode, e.g.

```python
with torch.autograd.set_deterministic_execution_order(True):
    loss.backward()
```

with the following contract:

1. Only valid with `set_multithreading_enabled(False)` and a single-device graph (error otherwise, at least initially).
2. The engine executes nodes in exactly the order `get_current_graph_task_execution_order()` returns, rather than popping the ready queue opportunistically. By construction that order respects dependencies, so no deadlock is possible; the change is to *wait* for the scheduled next node instead of running whatever is ready.
3. Consequence for `saved_tensors_hooks`: unpack requests arrive in a deterministic, documented order (the predicted schedule), which for single-threaded forward corresponds to reverse forward execution where dependencies allow.

This does not change the default. The default engine behavior and the "ordering is unspecified" contract from #198016 stay as they are.

### Known costs and open questions

- Precomputing the schedule requires an upfront graph traversal. `compute_dependencies` already traverses the graph, so the extra cost is one heap-ordered pass; acceptable for an opt-in mode, but worth measuring.
- In strict order, the engine may idle where it previously did useful work on another branch. For the target workloads (memory-bound, single device) this is an acceptable trade; numbers to be provided with the PR.
- Reentrant backward and recompute-in-backward (e.g. `torch.utils.checkpoint`) create nodes during execution that are not in the precomputed schedule. Options: (a) forbid in strict mode, (b) treat the recomputed subgraph as a nested strict schedule. Input welcome.
- `sequence_nr` collisions from multi-threaded forward make the schedule ill-defined; hence restriction 1.

### Alternatives considered

- Documenting the status quo — done in #198016; it clarifies but does not help sequential-reconstruction hooks.
- Defensive library code — works (awslabs/keys_values#149) but duplicates effort and memory in every library with this pattern.
- Prefetching all saved tensors for a region before backward — defeats the purpose of memory-reduction hooks.

I am happy to implement this behind the flag, with benchmarks, if the approach (or an amended one) is acceptable to the autograd maintainers.

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93
